File: android/app/src/main/java/com/imatrixteam/jsflutter/MXJSFlutterEngine.java
//  MXFlutterFramework
//  Copyright 2019 The MXFlutter Authors. All rights reserved.
//
//  Use of this source code is governed by a MIT-style license that can be
//  found in the LICENSE file.

package com.imatrixteam.jsflutter;

import android.content.Context;

import com.imatrixteam.jsflutter.utils.FileUtils;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;

import io.flutter.Log;
import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

import static com.imatrixteam.jsflutter.MXJSFlutterApp.JSFLUTTER_SRC_DIR1;

public class MXJSFlutterEngine {

    public static final String TAG = "MXJSFlutterEngine";

    private String rootPath;
    private MXJSFlutterApp currentApp;

    private Context mContext;
    private BinaryMessenger mFlutterEngine;

    private MXJSEngine mJsEngine;

    //Flutter通道
    private static final String FLUTTER_METHED_CHANNEL_NAME = "js_flutter.flutter_main_channel";
    MethodChannel jsFlutterAppChannel;

    private boolean isFlutterEngineIsDidRender;
    private ArrayList<MethodCall> callFlutterQueue;

    public MXJSFlutterEngine(Context context, BinaryMessenger flutterEngine) {
        this.mContext = context;
        this.mFlutterEngine = flutterEngine;
        setup();
    }

    public void setJsEngine(MXJSEngine jsEngine) {
        mJsEngine = jsEngine;
    }

    public void setup() {
        rootPath = JSFLUTTER_SRC_DIR1;
        setupChannel();
//        callFlutterQueue = new ArrayList<>(1);
//        isFlutterEngineIsDidRender = true;
//        for (MethodCall call : callFlutterQueue) {
//            jsFlutterAppChannel.invokeMethod(call.method, call.arguments);
//        }
    }

    public void setupChannel() {
        jsFlutterAppChannel = new MethodChannel(mFlutterEngine, FLUTTER_METHED_CHANNEL_NAME);
        jsFlutterAppChannel.setMethodCallHandler(new MethodChannel.MethodCallHandler() {
            @Override
            public void onMethodCall(MethodCall methodCall, MethodChannel.Result result) {
                if (methodCall.method.equals("callNativeRunJSApp")) {
                    String jsAppName = methodCall.<String>argument("jsAppName");
                    runApp(jsAppName);
                    result.success("success");
                } else if (methodCall.method.equals("callJsCallbackFunction")) {
                    String jsAppName = methodCall.<String>argument("callbackId");
                    String param = methodCall.<String>argument("param");
                    callJSCallbackFunction(jsAppName, param);
                    result.success("success");
                } else if (methodCall.method.equals("mxLog")) {
                    Log.i(TAG, methodCall.arguments.toString());
                }
            }
        });
    }


    public void destroy() {
        if (currentApp != null) {
            currentApp.close();
        }
    }

    public boolean showPage(String pageName) {
        return true;
    }

    public void runApp(String appName) {
        if (currentApp != null) {
            currentApp.close();
            currentApp = null;
        }

        JSModule.clearModuleCache();

        currentApp = new MXJSFlutterApp();
        currentApp.initWithAppName(mContext, rootPath + "/" + appName, rootPath, this);
        currentApp.runAppWithPageName();
    }

    public void callJSCallbackFunction(String callbackId, String params) {
        //todo
    }

    public void callFlutterReloadAppWithJSWidgetData(String widgetData) {
        callFlutterReloadAppWithRouteName("MXJSWidget", widgetData);
    }

    public void callFlutterReloadAppWithRouteName(String routeName, String widgetData) {
        FileUtils.assertUiThread();

        if (routeName == null || widgetData == null) {
            return;
        }
        try {
            JSONObject jsonObject = new JSONObject();
            jsonObject.put("routeName", routeName);
            jsonObject.put("widgetData", widgetData);
            MethodCall call = new MethodCall("reloadApp", jsonObject.toString());
//            if (isFlutterEngineIsDidRender) {
//                callFlutterQueue.add(call);
//                return;
//            }
            jsFlutterAppChannel.invokeMethod(call.method, call.arguments);
        } catch (JSONException e) {
            Log.e(TAG, "", e);
        }
    }

    public void callFlutterMethodChannelInvoke(String channelName, String methodName, JSONObject params, MethodChannel.Result callback) {
        FileUtils.assertUiThread();

        try {
            JSONObject jsonObject = new JSONObject();
            jsonObject.put("channelName", channelName);
            jsonObject.put("methodName", methodName);
            jsonObject.put("params", params);
            MethodCall call = new MethodCall("mxflutterBridgeMethodChannelInvoke", jsonObject.toString());
            jsFlutterAppChannel.invokeMethod(call.method, call.arguments, callback);
        } catch (JSONException e) {
            Log.e(TAG, "", e);
        }
    }

    public void callFlutterEventChannelReceiveBroadcastStreamListenInvoke() {
        FileUtils.assertUiThread();

        //todo
    }
}

File: ios/MXFlutterOCFramework/MXJSFlutterEngine.m
//
//  MXJSFlutterEngine.m
//  Runner
//
//  Created by soapyang on 2018/12/22.
//  Copyright © 2018 The Chromium Authors. All rights reserved.
//
//  Use of this source code is governed by a MIT-style license that can be
//  found in the LICENSE file.

#import "MXJSFlutterEngine.h"

#import "JSModule.h"
#import "MXJSFlutterDefines.h"
#import <Flutter/Flutter.h>
#import "AppDelegate.h"
#import "MXJSEngine.h"

@interface MXJSFlutterEngine ()
{

}

@property (nonatomic, strong) NSString *rootPath;


@property (nonatomic, strong) FlutterMethodChannel *basicChannel;
@property (nonatomic, strong) NSMutableArray<FlutterMethodCall*> *callFlutterQueue;

@end

@implementation MXJSFlutterEngine

- (instancetype)initWithEngine:(FlutterEngine*)engine
{
    if (self = [super init])
    {
        self.flutterEngine = engine;
        [self setup];
    }
    return self;
}

- (void)setup
{
    [self unsetup];
    
    self.rootPath = [JSFLUTTER_SRC_BASE_DIR stringByAppendingPathComponent:JSFLUTTER_SRC_DIR];
    self.callFlutterQueue = [NSMutableArray arrayWithCapacity:2];

    [self setupChannel];
}

- (void)setupChannel
{
    
    self.basicChannel = [FlutterMethodChannel
                         methodChannelWithName:@"js_flutter.flutter_main_channel"
                         binaryMessenger:self.flutterEngine.binaryMessenger];
    
    __weak MXJSFlutterEngine *weakSelf = self;
    [self.basicChannel setMethodCallHandler:^(FlutterMethodCall * _Nonnull call, FlutterResult  _Nonnull result) {
        __strong MXJSFlutterEngine *strongSelf = weakSelf;
        if (!strongSelf) {
            return;
        }
        
        if ([call.method isEqualToString:@"callNativeRunJSApp"]) {
            [strongSelf callNativeRunJSApp:call.arguments];
        } else if ([call.method isEqualToString:@"callJsCallbackFunction"]) {
            [strongSelf callJsCallBackFunction:call.arguments];
        }
        else if ([call.method isEqualToString:@"mxLog"]) {
            NSLog(@"%@", call.arguments);
        }
    }];
    
}

- (void)unsetup {

}

- (void)runJSApp:(NSString*)appName
{
    [self runApp:appName];
}

- (void)callJsCallBackFunction:(id)arguments
{
    NSDictionary *argsMap = arguments;
    NSString *callbackId = argsMap[@"callbackId"];
    NSString *param = argsMap[@"param"];
    
    [self.jsEngine callJSCallbackFunction:callbackId param:param];
}

//MARK: - flutter -> Native
//由Flutter 代码启动JSApp。 可以用在先显示Dart页面，然后路由调转到JS页面
//启动JSApp之后，执行JS代码，JS代码可以主动调用Flutter显示自己的页面，也能接受Flutter的指令，显示对应页面
- (void)callNativeRunJSApp:(id)arguments
{
    NSDictionary *argsMap = arguments;
    NSString *jsAppName = argsMap[@"jsAppName"];

    [self runApp:jsAppName];
}

//MARK: - native -> flutter
//--------------------------------------------
//调用Flutter切换到JSWidget，显示JS渲染的内容
//MARK: - native -> flutter
//--------------------------------------------
- (void)callFlutterReloadAppWithJSWidgetData:(NSString*)widgetData
{
    [self callFlutterReloadAppWithRouteName:@"MXJSWidget" widgetData:widgetData];
}

- (void)callFlutterReloadAppWithRouteName:(NSString*)routeName widgetData:(NSString*)widgetData
{
    routeName = routeName?routeName:@"";
    widgetData = widgetData?widgetData:@"";
    
    NSMutableDictionary *arguments = [[NSMutableDictionary alloc] init];
    if (routeName)
    {
        [arguments setObject:routeName forKey:@"routeName"];
    }
    if (widgetData)
    {
        [arguments setObject:widgetData forKey:@"widgetData"];
    }
    FlutterMethodCall* call  = [FlutterMethodCall methodCallWithMethodName:@"reloadApp" arguments:arguments];
    
//    if (!_flutterEngineIsDidRender) {
//
//        [self.callFlutterQueue addObject:call];
//        return;
//    }
    
    [self.basicChannel invokeMethod:call.method arguments:call.arguments];
}

- (void)callFlutterMethodChannelInvoke:(NSString*)channelName methodName:(NSString*)methodName params:(NSDictionary *)params callback:(void(^)(id _Nullable result))callback
{
    NSMutableDictionary *arguments = [[NSMutableDictionary alloc] init];
    if (channelName)
    {
        [arguments setObject:channelName forKey:@"channelName"];
    }
    if (methodName)
    {
        [arguments setObject:methodName forKey:@"methodName"];
    }
    if (params)
    {
        [arguments setObject:params forKey:@"params"];
    }
    FlutterMethodCall* call = [FlutterMethodCall methodCallWithMethodName:@"mxflutterBridgeMethodChannelInvoke" arguments:arguments];
//    if (!_flutterEngineIsDidRender) {
//        [self.callFlutterQueue addObject:call];
//        return;
//    }
    
    [self.basicChannel invokeMethod:call.method arguments:call.arguments result:^(id  _Nullable result) {
        if (callback) {
            callback(result);
        }
    }];
}

- (void)callFlutterEventChannelReceiveBroadcastStreamListenInvoke:(NSString*)channelName
                                                      streamParam:(NSString *)streamParam
                                                         onDataId:(NSString *)onDataId
                                                        onErrorId:(NSString *)onErrorId
                                                         onDoneId:(NSString *)onDoneId
                                                    cancelOnError:(NSNumber *)cancelOnError
                            
{
    if ([streamParam isEqualToString:@"null"]) {
        streamParam = nil;
    }
    
    NSMutableDictionary *arguments = [[NSMutableDictionary alloc] init];
    if (channelName)
    {
        [arguments setObject:channelName forKey:@"channelName"];
    }
    if (streamParam)
    {
        [arguments setObject:streamParam forKey:@"streamParam"];
    }
    if (onDataId)
    {
        [arguments setObject:onDataId forKey:@"onDataId"];
    }
    if (onErrorId)
    {
        [arguments setObject:onErrorId forKey:@"onErrorId"];
    }
    if (onErrorId)
    {
        [arguments setObject:onErrorId forKey:@"onErrorId"];
    }
    if (cancelOnError)
    {
        [arguments setObject:cancelOnError forKey:@"cancelOnError"];
    }
    FlutterMethodCall* call = [FlutterMethodCall methodCallWithMethodName:@"mxflutterBridgeEventChannelReceiveBroadcastStreamListenInvoke"
                                                                arguments:arguments];
    
    [self.basicChannel invokeMethod:call.method arguments:call.arguments];
}


- (void)runApp:(NSString*)appName 
{
    //退出原来的APP
    if (self.currentApp) {
        [self.currentApp exitApp];
        self.currentApp = nil;
    }
    
    [JSModule clearModuleCache];
    NSString *appRootPath = [self.rootPath stringByAppendingPathComponent:appName];
    self.currentApp  = [[MXJSFlutterApp alloc] initWithAppName:appName engine:self appRootPath:appRootPath];
    
    [self.currentApp runApp];
}



@end
